### ISSUE TITLE
Compile error when trying to use GenericDocument::Parse with a different source encoding

### ISSUE THREAD

**Issue by Leucotheia** [miloyip/rapidjson#427]
Hi, while attempting to parse a utf-8 encoded string and create a utf-16 encoded GenericDocument I noticed that the parameter type of the Parse function is potentially incorrect. Specifically the type specified is that defined within the document rather than that of the source encoding as defined in the function template parameters.

I have made a quick fix locally and tested in my circumstances and everything seems fine, but I could of course be missing something more subtle. Diff of my change is below, but an alternate workaround I guess would be to create a StringStream using the source encoding and pass that instead.

```
 include/rapidjson/document.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/include/rapidjson/document.h b/include/rapidjson/document.h
index f5c1be9..a42b47f 100644
--- a/include/rapidjson/document.h
+++ b/include/rapidjson/document.h
@@ -1951,7 +1951,7 @@ public:
         \param str Read-only zero-terminated string to be parsed.
     */
     template <unsigned parseFlags, typename SourceEncoding>
-    GenericDocument& Parse(const Ch* str) {
+	GenericDocument& Parse( const typename SourceEncoding::Ch* str ) {
         RAPIDJSON_ASSERT(!(parseFlags & kParseInsituFlag));
         GenericStringStream<SourceEncoding> s(str);
         return ParseStream<parseFlags, SourceEncoding>(s);
```

**Comment by miloyip** [comment]
I think you are correct. It looks like a bug.
Would you mind submitting a PR with this? (Adding an test case would be gerat.)

**Comment by Leucotheia** [comment]
Will do. I'm in the middle of something else at the moment but I will get to it this week.  Cheers.

**Comment by pah** [comment]
This has been fixed in #478, right?

**Comment by miloyip** [comment]
Yes.
